test/unit/context/test-node-components.js

```diff
@@ -0,0 +1,631 @@
+
+      beforeEach(() => {
+        component1Spy = sandbox.spy();
+        Component1 = withMetaData(component1Spy);
+
+        componentWithDepsSpy = sandbox.spy();
+        ComponentWithDeps = withMetaData(
+          [NonRecursive, Concat],
+          componentWithDepsSpy
+        );
+
+        componentWithCleanupSpy = sandbox.spy();
+        cleanupSpy = sandbox.spy();
+        ComponentWithCleanup = (...args) => {
+          componentWithCleanupSpy.apply(null, args);
+          return cleanupSpy;
+        };
+      });
+
+      it('should only call component once w/o input', () => {
+        setComponent(parent.node, Component1);
+        expect(component1Spy).to.not.be.called;
+
+        clock.runAll();
+        expect(component1Spy).to.be.calledOnce.calledWith(
+          parent.node,
+          undefined
+        );
+
+        setComponent(parent.node, Component1);
+        clock.runAll();
+        expect(component1Spy).to.be.calledOnce; // no changes.
+      });
+
+      it('should only call component once per input', () => {
+        setComponent(parent.node, Component1, 1);
+        clock.runAll();
+        expect(component1Spy).to.be.calledOnce.calledWith(parent.node, 1);
+
+        // Rerun the component due to the input change.
+        setComponent(parent.node, Component1, 2);
+        clock.runAll();
+        expect(component1Spy).to.be.calledTwice.calledWith(parent.node, 2);
+
+        // Input didn't change - do not rerun.
+        setComponent(parent.node, Component1, 2);
+        clock.runAll();
+        expect(component1Spy).to.be.calledTwice;
+      });
+
+      it('should reconnect component when the node is reconnected', async () => {
+        setComponent(parent.node, Component1, 1);
+        clock.runAll();
+        expect(component1Spy).to.be.calledOnce.calledWith(parent.node, 1);
+
+        parent.node.remove();
+        await rediscover(parent);
+        clock.runAll();
+
+        setComponent(parent.node, Component1, 2);
+        clock.runAll();
+        expect(component1Spy).to.be.calledOnce;
+
+        grandparent.node.appendChild(parent.node);
+        await rediscover(parent);
+        clock.runAll();
+        expect(component1Spy).to.be.calledTwice.calledWith(parent.node, 2);
+      });
+
+      it('should wait until all deps satisfied', () => {
+        setComponent(parent.node, ComponentWithDeps, 1);
+
+        clock.runAll();
+        expect(componentWithDepsSpy).to.not.be.called;
+
+        grandparent.values.set(Concat, 'OWNER1', 'A');
+        clock.runAll();
+        expect(componentWithDepsSpy).to.not.be.called;
+
+        parent.values.set(Concat, 'OWNER1', 'B');
+        clock.runAll();
+        expect(componentWithDepsSpy).to.not.be.called;
+
+        parent.values.set(NonRecursive, 'OWNER1', 'NR');
+        clock.runAll();
+        expect(componentWithDepsSpy).to.be.calledOnce.calledWith(
+          parent.node,
+          1,
+          'NR',
+          'AB'
```

Comment:
nit/FFTI: instead of just testing the component deps logic, this also ends up testing value propagation (that the concat behaved as expected)

Reply:
Partially, yes. I wanted this test b/c a concat-like property cascades through the tree. And if note careful, we could force multiple updates with intermediate values, when only one run would suffice.